Sources/SwiftProtobuf/ProtobufJSONEncoding.swift

```diff
@@ -21,16 +21,32 @@ struct ProtobufJSONEncodingVisitor: ProtobufVisitor {
     private var encoder = ProtobufJSONEncoder()
     var result: String {return encoder.result}
 
-    init() {}
+    private var nameResolver: (Int) -> String?
+    private var anyTypeURL: String?
+
+    init(message: ProtobufJSONMessageBase, anyTypeURL: String? = nil) throws {
+        self.nameResolver =
+            ProtoNameResolvers.jsonFieldNameResolver(for: message)
+        self.anyTypeURL = anyTypeURL
 
-    init(message: ProtobufJSONMessageBase) throws {
         try withAbstractVisitor {(visitor: inout ProtobufVisitor) in
             try message.traverse(visitor: &visitor)
         }
     }
 
     mutating func withAbstractVisitor(clause: (inout ProtobufVisitor) throws -> ()) throws {
         encoder.startObject()
+
+        // TODO: This is a bit of a hack that exists as a workaround to make the
+        // hand-written Any serialization work with the new design. We need to
+        // generate those WKTs instead of maintaining the hand-written ones,
+        // handle the special cases differently, and then remove this.
```

Comment:
The way that JSON for `Any` and the other well-known types have so far been handled here was the most challenging thing to get working with the new design, because of the handwritten protos and special cases—hence the hack above. I don't plan to let that hack live for very long, but I didn't want to overload this PR with other work.

To fix this for good, we need to get rid of the handwritten protos and generate them like everything else. The side effect of doing that is that we can no longer customize the JSON representation in the type itself, because even Swift extensions can't override behaviors. So instead, we'd have to do something like the [Java implementation](https://github.com/google/protobuf/blob/master/java/util/src/main/java/com/google/protobuf/util/JsonFormat.java#L564), which just handles those special cases within the JSON printer itself.

While I can definitely appreciate the desire to make things generalized and extensible, in this case it's added a lot of complexity to the API that would be good to clean up. The protobuf spec says [this finite set of specific types is given special JSON treatment](https://developers.google.com/protocol-buffers/docs/proto3#json) and nothing more, so it's *better* to be restrictive and limit customization because we don't want to (a) increase our maintenance burden and (b) make it possible for anyone to write types that arbitrarily don't conform to the spec.